### PR TITLE
Error on negative scale down value; update oldRS array after scaling; add logs when scaling down

### PR DESCRIPTION
**Pull request by janetkuo**
Addresses #21810 #22719 

@kubernetes/sig-config

### CODE REVIEW

**Comment by bgrant0607** [comment]
@janetkuo Was this problem reproducible enough to confirm whether this solves the problem?

Does it make sense to do this everywhere we call scaleReplicaSetAndRecordEvent (except maybe Recreate?)?

I agree that it seems like an improvement.

The next time the sync loop is invoked, the changes will be forgotten, until the state has been updated via watch, but I think the controller should make pretty much the same decisions again, and the updates should fail, anyway, if its state isn't up to date.

**Comment by bgrant0607** [comment]
LGTM

**Comment by janetkuo** [comment]
@bgrant0607 It's hard to reproduce. I've tried running this test locally (without this fix) but it passed 100 times. Adding logs will help debugging, if this fix doesn't solve the problem. Let's see if this happens again.

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -927,13 +930,17 @@ func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRSs []*extensions.Re
 
 		scaledDownCount := integer.IntMin(maxCleanupCount-totalScaledDown, targetRS.Spec.Replicas-readyPodCount)
 		newReplicasCount := targetRS.Spec.Replicas - scaledDownCount
-		_, _, err = dc.scaleReplicaSetAndRecordEvent(targetRS, newReplicasCount, deployment)
+		if newReplicasCount > targetRS.Spec.Replicas {
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Wouldn't it make sense to scale to zero?

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -927,13 +930,17 @@ func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRSs []*extensions.Re
...
+			return nil, 0, fmt.Errorf("when cleaning up unhealthy replicas, got invalid request to scale down %s/%s %d -> %d", targetRS.Namespace, targetRS.Name, targetRS.Spec.Replicas, newReplicasCount)
+		}
+		_, updatedOldRS, err := dc.scaleReplicaSetAndRecordEvent(targetRS, newReplicasCount, deployment)
 		if err != nil {
-			return totalScaledDown, err
+			return nil, totalScaledDown, err
 		}
 		totalScaledDown += scaledDownCount
+		oldRSs[i] = updatedOldRS
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Since this is a slice of pointers I would expect it to be updated w/o needing to re-assign

**Comment by eparis** [comment]
This PR is already included in the release-1.2 branch and does not require cherrypick-ing

**Pull request merged**